File: src/main/java/fr/centralesupelec/sio/data/MoviesRepository.java
package fr.centralesupelec.sio.data;

import fr.centralesupelec.sio.model.Movie;

import java.util.List;

public abstract class MoviesRepository {

    private static MoviesRepository sRepository;

    public static MoviesRepository getInstance() {
        if (sRepository == null) {
            sRepository = new DummyMoviesRepository();
        }
        return sRepository;
    }

    protected MoviesRepository() { }

    public abstract List<Movie> getMovies();

}

File: src/main/java/fr/centralesupelec/sio/data/MoviesRepository.kt
package fr.centralesupelec.sio.data

import fr.centralesupelec.sio.model.Movie

/**
 * A data repository to expose movie-related entities.
 */
interface MoviesRepository {

    companion object {
        val sharedInstance: MoviesRepository by lazy { DummyMoviesRepository() }
    }

    /**
     * Query all available movies.
     * @return A list of [Movie] entities.
     */
    fun getMovies(): List<Movie>

    /**
     * @param id: The id of the movie.
     * @return The [Movie] entity, or null of the movie does not exist.
     */
    fun getMovie(id: Long): Movie?

    // TODO: Add other movie-related methods here

}
